service/storage_service.cc

```diff
@@ -1825,8 +1825,60 @@ class topology_coordinator {
         slogger.info("raft topology: coordinator fiber found a node to work on id={} state={}", node.id, node.rs->state);
 
         switch (node.rs->state) {
-            case node_state::normal:
             case node_state::none: {
+                if (!node.rs->join_cookie) {
+                    on_internal_error(slogger, ::format(
+                            "Request to join node {} does not have the join cookie set",
+                            node.id));
+                }
+
+                // Check the parameters of the node to see whether we can add it
+                // to the cluster.
+                auto decision = validate_joining_node(node);
+
+                // If the validation failed, we will remove the request at the end.
+                bool should_drop = std::holds_alternative<join_node_response_params::rejected>(decision);
+                auto ip = _address_map.find(node.id);
+                if (!ip) {
+                    slogger.warn("raft topology: cannot send JOIN_NODE_RESPONSE to {} because mapping to ip is not available",
+                                 node.id);
+                    co_await coroutine::exception(std::make_exception_ptr(
+                            std::runtime_error(::format("no ip address mapping for {}", node.id))));
+                }
+                auto response = co_await ser::storage_service_rpc_verbs::send_join_node_response(
```

Comment:
Currently, the topology coordinator will not attempt to enable features if there are pending join requests (nodes in `none` state), so there is no risk.

If we wanted to first add the node to topology and only after that send join_node_response, then we need an additional step in the state machine (because we need to know whether we may need to send the response to the joining node).

Reply:
I was actually hoping that we change the procedure to add to topology first, then group 0 second.

Note that when removing a node, already do it this way: we first remove it from group 0, then "remove" it from topology (actually move it to "left" state). So I think we should do the reverse when joining the node; that way, a node's lifetime in group 0 will be bounded from above by the lifetime in topology, making things easier to reason about, and also making error recovery easier e.g. after a failed boot -- the coordinator could make decisions based just on topology state.

cc @gleb-cloudius

Reply:
I was distrusted from reviewing the series so did not get to here yet. but in the original proposal we added to topology before joining to the group0 IIRC.

Reply:
Now, the order for the accepted nodes is:

1. Move node to bootstrapping/replacing state
2. Add to group 0
3. Send join_node_response RPC